_posts/2015-09-12-developing-gamepad-virtual-keyboard.markdown

```diff
@@ -0,0 +1,150 @@
+
+ + Shows the cursor's location.
+ + Shows current text, this includes text that was already in the Real Input onFocus.
+ + Places new text directly after the cursor location and the cursor shifts one character to the right.
+ + Cursor can change position using the **mouse**, **keyboard**, OR the **gamepad**.
+ + Text can be entered with the **gamepad** OR **keyboard**.
+
+That first one is easy. The HTML input tag automatically shows the current location of the cursor.
+
+<p style=”text-align: center;”>
+<input placeholder=”this is an input”>
+<p>
+
+The 2nd one is almost as straight-forward. We’ll have to keep track of that current text string value. If you’re using Flux, you’ll probably want to create a Store for the value of the input tag at this point. That way our Fake Input and the Real Input will both mirror the same values when the virtual keyboard is dismissed.
+
+The functionality for those last 3 points all fold into each other, so we’ll discuss them all at once.
+
+Let’s assume you already have a mechanism for getting the currently selected symbol. We'll need a callback for that event, let's call it onGamepadEvent. Inside this callback we receive the new character being typed out. We can't just concatenate it onto the end of our current string because the cursor could be anywhere. So we need to get the current string value from the input, get the cursor location within that string, and then insert the new character at that location of the cursor. We also need to increment the location of the cursor, as I'll discuss next.
+
+```js
+// stores/input.js
+
+getValueDivision: function() {
+    var value = this.value + '';
+    var firstStrPart = value.substring(0, this.cursor),
+        secondStrPart = value.substring(this.cursor, value.length);
+
+    return {
+        start: firstStrPart,
+        end: secondStrPart
+    }
+},
+
+onGamepadEvent: function(symbol) {
+    var valueParts = this.getValueDivision();
+    this.value = valueParts.start + symbol + valueParts.end;
+    this.cursor++;
+    this.emit('change');
+},
+
+```
+
+Since the location of the cursor can be changed by the mouse, keyboard, and the gamepad it's impractical to listen for keyboard events and mouse clicks to track the cursor, and in turn, know where to add new characters. Instead, let's use the onChange event for the Fake input and update a value that we store to keep track of the cursor. The DOM provides a decent API to work with text selection and the cursor, called the [Selection API](https://developer.mozilla.org/en-US/docs/Web/API/Selection). You'll need polyfills to get support for IE, here are the ones for [getCursor](http://stackoverflow.com/a/263796/2687479) and [setCursor](http://stackoverflow.com/questions/1865563/set-cursor-at-a-length-of-14-onfocus-of-a-textbox/1867393#1867393)
```

Comment:
You'll need polyfills to get support for IE - here are the ones for [getCursor](http://stackoverflow.com/a/263796/2687479) and [setCursor](http://stackoverflow.com/questions/1865563/set-cursor-at-a-length-of-14-onfocus-of-a-textbox/1867393#1867393):